From 99b8c173daee735c6a8dcd7c939338f7d7b95c9b Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Mon, 11 Apr 2022 07:34:08 +0300
Subject: [PATCH] [SW-79498] gaudi3: rma: override sig val efuse bit

In post-production phase, in case a rma is detected in a device,
the security requirement is that a rollback to a previously stable
version should be allowed.

This effectively means we need to bypass the svn checks in case of
a rma detection.

As per the security requirement, the fw allows to run an unauthenticated
image if rma is set.

Change-Id: I83385f9126d6993170c259ea56a3b541259900c1
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 571a4b8..de1d0ab 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -371,10 +371,23 @@
 		arcboot_rom_status_update(PROGRESS_INIT_SPI_BOOT_DESC_FAIL);
 }
 
+#define BOOT_STRAP_RMA_MODE_MASK			0x3
+
 void arcboot_rom_fuse_override(void)
 {
+	uint32_t rma_mode_status;
+	uint32_t strap_h = regmap_readl(mmD0_PSOC_BOOT_CONF_BOOT_STRAP_PINS_H);
+
 	if (arcboot_rom_global.config_fuse & EFUSE_SYSCFG_EOM_MASK)
 		arcboot_rom_global.config_fuse |= EFUSE_SYSCFG_PRE_PRODUCTION_BITS;
+
+	rma_mode_status = (strap_h & D0_PSOC_BOOT_CONF_BOOT_STRAP_PINS_H_MODULE_ID_M)
+			>> D0_PSOC_BOOT_CONF_BOOT_STRAP_PINS_H_MODULE_ID_S;
+
+	rma_mode_status >>= BOOT_STRAP_RMA_MODE_MASK;
+	/* TODO: add a relavent gpio check for RMA mode once finalized for gaudi3 */
+	if ((rma_mode_status & BOOT_STRAP_RMA_MODE_MASK) == BOOT_STRAP_RMA_MODE_MASK)
+		arcboot_rom_global.config_fuse &= ~EFUSE_SYSCFG_ENABLE_VALIDATION_MASK;
 }
 
 void arcboot_rom_fuse(void)
